### DISCUSSION TITLE
Automatic check code on saving a java file (pmd-eclipse-plugin)

### DISCUSSION THREAD

**Discussion by caesarkim** [pmd/pmd#2713] [Q&A] [answered]
**Description:**
Is there a way to force/configure an automatic 'Check Code' when a java source file is saved?  I have my custom rule set which I want to be ran automatically against the Java source file which I am working on when I save the changes. 

Please let me know if there is already a configuration for this. If not, I appreciate if someone can provide any resource.

Thanks.

**Comment by jsotuyod** [comment]
@caesarkim PMD does not currently support this natively, but there are good workarounds…

Firstly, PMD's [incremental analysis](https://pmd.github.io/pmd-6.26.0/pmd_userdocs_incremental_analysis.html) allows it to only analyze modified files, which certainly helps the performance for such a use case. Enabling that would be strongly recommended.

What would be missing is the ability to watch the file system for file changes to trigger the PMD run.

Off the top of my mind, there are 2 approaches that could currently work…

1. If you are using Gradle, you can run it in [continous mode](https://docs.gradle.org/current/userguide/command_line_interface.html#sec:continuous_build) this will effectively rerun PMD upon changes (and even make sure the code is recompiled so type resolution works effectively!). If already using Gradle to build and run PMD this is effectively free!
2. Use a tool such as [Watchman](https://facebook.github.io/watchman/) or [Filewatcher](https://github.com/filewatcher/filewatcher) to detect changes, and run a custom analysis script. Such script would have to:
    1. Recompile changed files for best results
    2. Rerun PMD on the whole project (not changed files!), incremental analysis will work it out on it's own.

**Comment by adangel** [comment]
yes, PMD alone doesn't support this. But there are are many IDE plugins, e.g. for eclipse and intellij and others. These plugins usually hook into the build infra of the IDE and can trigger a rerun of static analysis tools when a file is changed.
So it depends which IDE you are using.

Eclipse Plugins:
* https://github.com/pmd/pmd-eclipse-plugin/
* https://github.com/eclipse-pmd/eclipse-pmd

IntelliJ Plugin
* https://plugins.jetbrains.com/plugin/4596-qaplug--pmd

And there are many other plugins/tools: https://pmd.github.io/latest/pmd_userdocs_tools.html

**Comment by caesarkim** [comment]
> yes, PMD alone doesn't support this. But there are are many IDE plugins, e.g. for eclipse and intellij and others. These plugins usually hook into the build infra of the IDE and can trigger a rerun of static analysis tools when a file is changed.
> So it depends which IDE you are using.
> 
> Eclipse Plugins:
> 
> * https://github.com/pmd/pmd-eclipse-plugin/
> * https://github.com/eclipse-pmd/eclipse-pmd
> 
> IntelliJ Plugin
> 
> * https://plugins.jetbrains.com/plugin/4596-qaplug--pmd
> 
> And there are many other plugins/tools: https://pmd.github.io/latest/pmd_userdocs_tools.html

Actually, i was using pmd-eclipse-plugin. I saw the check box 'Check code after saving' under Window -> Preferences -> PMD. I made some changes on the java source code, but it doesn't seem that cod analysis is triggered. I don't see any violation even though I see some when I manually do 'Check Code'. 

Am I missing anything?

**Comment by caesarkim** [comment]
> @caesarkim PMD does not currently support this natively, but there are good workarounds…
> 
> Firstly, PMD's [incremental analysis](https://pmd.github.io/pmd-6.26.0/pmd_userdocs_incremental_analysis.html) allows it to only analyze modified files, which certainly helps the performance for such a use case. Enabling that would be strongly recommended.
> 
> What would be missing is the ability to watch the file system for file changes to trigger the PMD run.
> 
> Off the top of my mind, there are 2 approaches that could currently work…
> 
> 1. If you are using Gradle, you can run it in [continous mode](https://docs.gradle.org/current/userguide/command_line_interface.html#sec:continuous_build) this will effectively rerun PMD upon changes (and even make sure the code is recompiled so type resolution works effectively!). If already using Gradle to build and run PMD this is effectively free!
> 2. Use a tool such as [Watchman](https://facebook.github.io/watchman/) or [Filewatcher](https://github.com/filewatcher/filewatcher) to detect changes, and run a custom analysis script. Such script would have to:
>    
>    1. Recompile changed files for best results
>    2. Rerun PMD on the whole project (not changed files!), incremental analysis will work it out on it's own.

I am sorry that I should made my question a little clear. I was using pmd-eclipse plugin in Eclipse. I want to automatically trigger 'Check Code' when a change is made on the java source file and the java file is successfully compiled.

**Comment by adangel** [comment]
> Actually, i was using pmd-eclipse-plugin. I saw the check box 'Check code after saving' under Window -> Preferences -> PMD. I made some changes on the java source code, but it doesn't seem that cod analysis is triggered. I don't see any violation even though I see some when I manually do 'Check Code'.

Do you have "Project -> Build Automatically" enabled? Do you have in the project properties of your project under "PMD" the option "Enable PMD" enabled?

**Comment by adangel** [comment, marked as answer]
Since Version 4.0.16 the following option controls the automatic check:

> * PMD is now only executed automatically, if the option "Check code after saving" is enabled. This allows to simply disable automatic PMD checks temporarily. Executing PMD via the project's context menu "Check Code" is not affected and can still be used to manually execute PMD to update the markers.
>
> (https://github.com/pmd/pmd-eclipse-plugin/blob/master/ReleaseNotes.md#12-april-2018-4016v20180412-0833)

If the option "Check code after saving" is enabled, then:
* if "Build Automatically" is enabled, PMD is executed via the project builder
* if "Build Automatically" is disabled, PMD is executed via a resource change listener. Note: you might not get accurate results, since the project is not built completely (only the file you just changed, but not dependent files) and thus type resolution might not work correctly.

In both cases, PMD is only executed, if PMD is enabled in the project properties.

If PMD is not enabled in the project, the only way to execute PMD is manual via the project's context menu "Check Code".
